windows/hid.c

```diff
@@ -409,6 +409,98 @@ static void* hid_internal_get_device_interface_property(const wchar_t* interface
 	return property_value;
 }
 
+static int hid_internal_extract_int_token_value(const wchar_t* string, const wchar_t* token)
+{
+	int token_value;
+	wchar_t* startptr, * endptr;
+
+	startptr = wcsstr(string, token);
+	if (!startptr)
+		return -1;
+
+	startptr += wcslen(token);
+	token_value = wcstol(startptr, &endptr, 16);
+	if (endptr == startptr)
+		return -1;
+
+	return token_value;
+}
+
+static void hid_internal_get_usb_info(struct hid_device_info* dev, DEVINST dev_node)
+{
+	wchar_t *device_id = NULL, *hardware_ids = NULL;
+
+	device_id = hid_internal_get_devnode_property(dev_node, &DEVPKEY_Device_InstanceId, DEVPROP_TYPE_STRING);
+	if (!device_id)
+		goto end;
+
+	/* Check for Xbox Common Controller class (XUSB) device.
+	   https://docs.microsoft.com/windows/win32/xinput/directinput-and-xusb-devices
+	   https://docs.microsoft.com/windows/win32/xinput/xinput-and-directinput
+	*/
+	if (hid_internal_extract_int_token_value(device_id, L"IG_") != -1) {
+		/* Get devnode parent to reach out USB device. */
+		if (CM_Get_Parent(&dev_node, dev_node, 0) != CR_SUCCESS)
+			goto end;
+	}
+
+	/* Get the hardware ids from devnode */
+	hardware_ids = hid_internal_get_devnode_property(dev_node, &DEVPKEY_Device_HardwareIds, DEVPROP_TYPE_STRING_LIST);
+	if (!hardware_ids)
+		goto end;
+
+	for (wchar_t* hardware_id = hardware_ids; *hardware_id; hardware_id += wcslen(hardware_id) + 1) {
+		/* Normalize to upper case */
+		for (wchar_t* p = hardware_id; *p; ++p) *p = towupper(*p);
+
+		/* USB Device Interface Number.
+		   It can be parsed out of the Hardware ID if a USB device is has multiple interfaces (composite device).
+		   https://docs.microsoft.com/windows-hardware/drivers/install/standard-usb-identifiers
+		   https://docs.microsoft.com/windows-hardware/drivers/usbcon/enumeration-of-interfaces-not-grouped-in-collections
+		*/
+		dev->interface_number = hid_internal_extract_int_token_value(hardware_id, L"MI_");
```

Comment:
I have a game controller that has only one interface and `&MI_` is not present in none of its HardwareIDs.
At the end we end-up setting `interface_number` to `-1` for that device.
Do you thing we could add euristic to default it to `0` for devices with only one interface?
I didn't give it enough thought as of how to do so correctly.

Reply:
My device:
```
        +++++++++++++++++ Device Information ++++++++++++++++++
Device Description       : Xbox 360 Controller for Windows
Device Path 1            : \\?\USB#VID_045E&PID_028E#6&328b81da&1&2#{a5dcbf10-6530-11d2-901f-00c04fb951ed} (GUID_DEVINTERFACE_USB_DEVICE)
Device Path 2            : \\?\USB#VID_045E&PID_028E#6&328b81da&1&2#{ec87f1e3-c13b-4100-b5f7-8b84d54260cb}
Kernel Name              : \Device\USBPDO-9
Device ID                : USB\VID_045E&PID_028E\6&328B81DA&1&2
Hardware IDs             : USB\VID_045E&PID_028E&REV_2020 USB\VID_045E&PID_028E
Driver KeyName           : {d61ca365-5af4-4486-998b-9db4734c6ca3}\0001
Driver                   : \SystemRoot\System32\drivers\xusb22.sys (Version: 10.0.22000.653  Date: 2022-05-13)
Driver Inf               : C:\WINDOWS\inf\xusb22.inf
Legacy BusType           : PNPBus
Class                    : XnaComposite
Class GUID               : {d61ca365-5af4-4486-998b-9db4734c6ca3}
Service                  : xusb22
Enumerator               : USB
Location Info            : Port_#0002.Hub_#0008
Location IDs             : PCIROOT(0)#PCI(1400)#USBROOT(0)#USB(10)#USB(2), ACPI(_SB_)#ACPI(PC00)#ACPI(XHCI)#ACPI(RHUB)#ACPI(HS10)#USB(2)
Container ID             : {d12ad581-3b3d-11ed-9df4-40b076a530b2}
Manufacturer Info        : Microsoft
Capabilities             : 0x84 (Removable, SurpriseRemovalOK)
Status                   : 0x0180600A (DN_DRIVER_LOADED, DN_STARTED, DN_DISABLEABLE, DN_REMOVABLE, DN_NT_ENUMERATOR, DN_NT_DRIVER)
Problem Code             : 0
HcDisableSelectiveSuspend: 0
EnableSelectiveSuspend   : 0
SelectiveSuspendEnabled  : 0
EnhancedPowerMgmtEnabled : 0
IdleInWorkingState       : 0
WakeFromSleepState       : 0
Power State              : D0 (supported: D0, D2, D3, wake from D0, wake from D2)
 Child Device 1          : USB Input Device
  Device ID              : USB\VID_045E&PID_028E&IG_00\7&C002879&0&00
  Class                  : HIDClass
  Driver KeyName         : {745a17a0-74d3-11d0-b6fe-00a0c90f57da}\0124 (GUID_DEVCLASS_HIDCLASS)
  Service                : HidUsb
  Location               : XBOX_360_DEVICE_00:00
   Child Device 1        : HID-compliant game controller
    Device Path          : \\?\HID#VID_045E&PID_028E&IG_00#8&3e51efe&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030} (GUID_DEVINTERFACE_HID)
    Kernel Name          : \Device\00000156
    Device ID            : HID\VID_045E&PID_028E&IG_00\8&3E51EFE&0&0000
    Class                : HIDClass
    Driver KeyName       : {745a17a0-74d3-11d0-b6fe-00a0c90f57da}\0125 (GUID_DEVCLASS_HIDCLASS)

    ---------------------- Device Descriptor ----------------------
bLength                  : 0x12 (18 bytes)
bDescriptorType          : 0x01 (Device Descriptor)
bcdUSB                   : 0x200 (USB Version 2.0) -> but device is Full-Speed only
bDeviceClass             : 0xFF (Vendor Specific)
bDeviceSubClass          : 0xFF
bDeviceProtocol          : 0xFF
bMaxPacketSize0          : 0x08 (8 bytes)
idVendor                 : 0x045E (Microsoft Corporation)
idProduct                : 0x028E
bcdDevice                : 0x2020
iManufacturer            : 0x01 (String Descriptor 1)
 *!*ERROR  String descriptor not found
iProduct                 : 0x02 (String Descriptor 2)
 Language 0x0409         : "XEOX Gamepad"
iSerialNumber            : 0x03 (String Descriptor 3)
 *!*ERROR  String descriptor not found
bNumConfigurations       : 0x01 (1 Configuration)
Data (HexDump)           : 12 01 00 02 FF FF FF 08 5E 04 8E 02 20 20 01 02   ........^...  ..
                           03 01    
```

Reply:
>Device ID              : USB\VID_045E&PID_028E&IG_00\7&C002879&0&00
Seems this device haven't reported proper serial number it its `iSerialNumber` string descriptor.

Reply:
> I have a game controller that has only one interface and `&MI_` is not present in none of its HardwareIDs. At the end we end-up setting `interface_number` to `-1` for that device. Do you think we could add euristic to default it to `0` for devices with only one interface? I didn't give it enough thought as of how to do so correctly.

 I think we should set `interface_number` to `0` for single-interface USB devices since this value is reported in its [USB_INTERFACE_DESCRIPTOR.iInterface](https://learn.microsoft.com/windows-hardware/drivers/ddi/usbspec/ns-usbspec-_usb_interface_descriptor) in this case.